
## BUSA8001- Programming Task 1  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written answer provide your solution in Markdown in the cells under each question.
- Comment out your print statements unless you are explicitly asked to print your output. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.

- In `application_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day (0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)

**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Provide your answers with print() and state them in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Provide your answers with print() and state them in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there now in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)

In [533]:
import pandas as pd
# 1. Load the CSV files into pandas DataFrames
df_application = pd.read_csv('/Users/albina/Documents/Macquarie Uni/Sem 3/Applied Predictive Analytics/Assignment1-S2-2024/data/application_record.csv')
df_credit = pd.read_csv('/Users/albina/Documents/Macquarie Uni/Sem 3/Applied Predictive Analytics/Assignment1-S2-2024/data/credit_record.csv')

In [534]:
# 2. Number of rows in df_application
rows_application = len(df_application)
print("Number of rows in df_application:", rows_application)

# Number of rows in df_credit
rows_credit = len(df_credit)
print("Number of rows in df_credit:", rows_credit)

Number of rows in df_application: 438445
Number of rows in df_credit: 1047185


# 2. Ans
There are 438445 number of rows in df_application.
There are 1047185 number of rows in df_credit.

In [535]:
#3. Number of unique clients in df_application
unique_clients_application = df_application['ID'].nunique()
print("Number of unique clients in df_application:", unique_clients_application)

# Number of unique clients in df_credit
unique_clients_credit = df_credit['ID'].nunique()
print("Number of unique clients in df_credit:", unique_clients_credit)

Number of unique clients in df_application: 438398
Number of unique clients in df_credit: 45924


# 3. Ans
There are 438398 number of unique bank clients in df_application.

There are 45924 number of unique bank clients in df_credit.

In [536]:
# 4. 

# Merging the two DataFrames
df = pd.merge(df_application, df_credit, on='ID', how='inner')

In [537]:
# 5. Number of rows in df
num_rows = len(df)
print("Number of rows in df:", num_rows)

# Number of unique clients in the merged DataFrame df
unique_clients = df['ID'].nunique()
print("Number of unique clients in df:", unique_clients)

Number of rows in df: 776325
Number of unique clients in df: 36396


# 5. Ans
There are 776325 number of rows in df.

There are 36396 number of uniquie clients in df.

# 6. How are the multiple rows for each ID in  df different ?

# Ans : 
The Multiple rows for each ID in df different because of following reasons:

While merging 'df_application' and 'df_credit', each rows in 'df' represents a combination of client information from 'df_application' and monthly credit status from 'df_credit'. 

The pd.merge() function merges two Dataframes based on specified column here i.e. ID which is common in both 'df_application' and df_credit'.

For every ID that exists in both DataFrames, the corresponding rows from both df_application and df_credit are combined into a single row in the resulting DataFrame df with the use of parrameter to inner.

Rows Excluded: Any 'ID' that is present in only one of the Dataframes (either 'df_application' or 'df_credit') is excluded from the result. This ensures that df contains only complete information about clients who have both demographic data and credit history.




---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)

2. Create a new *numpy* array called `list_of_past_due` that includes the unique ID numbers of clients whose `STATUS = 1` at any point during the last 12 months (hint: count the current month as the first month). (2 marks) 

3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` is in `list_of_past_due`, keeping only one row for each `ID` (hint: keep the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (hint: find out about `isin()` function in pandas.) (2 marks)

4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)

5. Increase `df_final` to a total of 4,500 rows by adding rows from `df` with unique `ID`s which are not in `list_of_past_due`. To do this start adding the rows from the beginning of `df`. (hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 

6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. (1 mark)

(10 marks)

In [538]:
#1 
# Define the mapping for STATUS values
status_mapping = {'C': 0, 'X': 0, '0': 0, '1': 1, '2': 1, '3': 1, '4': 1, '5': 1}

# Apply the mapping to the STATUS column in df
df['STATUS'] = df['STATUS'].map(status_mapping)

# Convert the STATUS column to integers to ensure they are encoded as integers
df['STATUS'] = df['STATUS'].astype(int)

In [539]:
#2
import numpy as np

# Filter rows where STATUS is 1 and MONTHS_BALANCE is within the last 12 months (0 to -11)
past_due_clients = df[(df['STATUS'] == 1) & (df['MONTHS_BALANCE'] >= -11)]

# Extract unique IDs from the filtered rows
list_of_past_due = np.unique(past_due_clients['ID'])


In [540]:
#3.  Create df_final by filtering rows in df where the ID is in list_of_past_due
df_final = df[df['ID'].isin(list_of_past_due)]

# Keep only the first occurrence of each ID
df_final = df_final.drop_duplicates(subset=['ID'], keep='first')

# Display the number of rows in df_final
print("Number of rows in df_final:", len(df_final))

# Display the first few rows of df_final to confirm the result
print(df_final.head())

Number of rows in df_final: 1737
           ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
369   5008831           F            N               Y           NaN   
954   5008872           M            Y               Y           NaN   
1718  5008912           F            N               Y           0.0   
1755  5008914           F            N               Y           0.0   
1790  5008916           F            N               Y           0.0   

      AMT_INCOME      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
369     157500.0               Working  Secondary / secondary special   
954     360000.0  Commercial associate                            NaN   
1718    297000.0  Commercial associate                            NaN   
1755    297000.0  Commercial associate                            NaN   
1790    297000.0  Commercial associate                            NaN   

        NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
369                

# 3. Ans

The code filters rows from a DataFrame df where the ID matches entries in list_of_past_due, then removes duplicate IDs, keeping only the first occurrence. The resulting DataFrame, df_final, contains 1,737 rows with unique IDs. The code also outputs the number of rows and displays a preview of the filtered data, which includes client demographic and financial information.

In [541]:
# 4.
# Add a new column y with value 1 for all rows
df_final['y'] = 1

In [542]:
#5 Find IDs that are not in list_of_past_due
non_past_due = df[~df['ID'].isin(list_of_past_due)]

# Keep adding rows until df_final has 4,500 rows
rows_to_add = 4500 - len(df_final)
additional_rows = non_past_due.drop_duplicates(subset=['ID']).iloc[:rows_to_add]

# Append these rows to df_final
df_final = pd.concat([df_final, additional_rows], ignore_index=True)

In [543]:
# 6.
#  Fill missing values in y with 0
df_final['y'] = df_final['y'].fillna(0)

# Remove STATUS and MONTHS_BALANCE columns
df_final = df_final.drop(['STATUS', 'MONTHS_BALANCE'], axis=1)

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
1. Delete `ID` column from `df_final` and reset its index. (1 marks)
2. Assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable in `df_final`, which variables are numeric and which ones are nominal? Answer this question by copying and completing the following table (6 marks)

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|||
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|||

3. Using appropriate functions find and comment on the missing values in `df_final` (3 marks)   

(10 marks)

In [544]:
#1 Delete the ID column from df_final and reset its index
df_final = df_final.drop(columns=['ID']).reset_index(drop=True)

# Q 3 .2 



| Variable Type | Number of Features | Features' List                                |
|---------------|--------------------|-----------------------------------------------|
| Numeric       | 5                  | CNT_CHILDREN, AMT_INCOME, DAYS_BIRTH, DAYS_EMPLOYED, CNT_FAM_MEMBERS|
| Ordinal       | 1                  | NAME_EDUCATION_TYPE                          |
| Nominal       | 12                 | CODE_GENDER, FLAG_OWN_CAR, FLAG_OWN_REALITY, NAME_INCOME_TYPE, NAME_FAMILY_STATUS, NAME_HOUSING_TYPE, FLAG_MOBIL, FLAG_WORK_PHONE, FLAG_PHONE, FLAG_EMAIL, OCCUPATION_TYPE, y |

In [545]:
#3
# Check for missing values in df_final
missing_values = df_final.isnull().sum()

# Display the missing values
print("Missing values in df_final:\n", missing_values)

# Comment on the missing values
print("\nComment:")
print("The output shows the count of missing values in each column of df_final. Columns with significant missing data may need imputation or removal, depending on their relevance.")


Missing values in df_final:
 CODE_GENDER               0
FLAG_OWN_CAR              0
FLAG_OWN_REALTY           0
CNT_CHILDREN             74
AMT_INCOME                0
NAME_INCOME_TYPE          0
NAME_EDUCATION_TYPE    1831
NAME_FAMILY_STATUS        0
NAME_HOUSING_TYPE         0
DAYS_BIRTH                0
DAYS_EMPLOYED             0
FLAG_MOBIL                0
FLAG_WORK_PHONE           0
FLAG_PHONE                0
FLAG_EMAIL                0
OCCUPATION_TYPE        1354
CNT_FAM_MEMBERS           0
y                         0
dtype: int64

Comment:
The output shows the count of missing values in each column of df_final. Columns with significant missing data may need imputation or removal, depending on their relevance.


---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
1. Use an appropriate `pandas` function to impute missing values in `df_final` (15 marks)
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes

(Total: 15 marks)

In [546]:
# Define a function to impute missing values based on variable type
def impute_missing_values(df):
    for column in df.columns:
        # If the column is numeric, fill missing values with the mean
        if df[column].dtype in ['int64', 'float64']:
            df[column] = df[column].fillna(df[column].mean())
        # If the column is categorical (nominal/ordinal), fill missing values with the mode
        else:
            df[column] = df[column].fillna(df[column].mode()[0])
    return df

# Impute missing values in df_final
df_final = impute_missing_values(df_final)

 
 ## Mean/Median/Mode Interpolation    
 
 In our data set we have encountered situtations where *categorical* and *numerical* data is missing   
 - Mean/Median/Mode Interpolation: Replace missing values with the mean, median, or mode of the column
 - However we have to be careful about the type of data we have
     - **Numeric data** -> I have used mean (if concerned about outliers or when the distribution is not symmetrical)    
     - **Categorical data** -> I have used mode (the only value that can be used with nominal data that can't be ordered)  
 

<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [555]:
# Question 5
# Define the mapping for NAME_EDUCATION_TYPE values
education_mapping = {
    'Lower secondary': 1,
    'Secondary / secondary special': 2,
    'Incomplete higher': 3,
    'Higher education': 4
}

# Apply the mapping to the NAME_EDUCATION_TYPE column
df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].replace(education_mapping).astype(int)

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [548]:
import pandas as pd

# Select nominal columns based on the given table
nominal_columns = ['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 
                   'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 
                   'OCCUPATION_TYPE']

# Create dummy variables and drop the first column of each set to avoid multicollinearity
df_final = pd.get_dummies(df_final, columns=nominal_columns, drop_first=True)

# Display the first few rows of the modified DataFrame to confirm the changes
print(df_final.head())

   CNT_CHILDREN  AMT_INCOME  NAME_EDUCATION_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0      0.403525    157500.0                    2      -10031          -1469   
1      0.403525    360000.0                    2      -16670          -5364   
2      0.000000    297000.0                    2      -15519          -3234   
3      0.000000    297000.0                    2      -15519          -3234   
4      0.000000    297000.0                    2      -15519          -3234   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL  CNT_FAM_MEMBERS  ...  \
0           1                0           1           0                2  ...   
1           1                0           1           0                2  ...   
2           1                0           0           0                1  ...   
3           1                0           0           0                1  ...   
4           1                0           0           0                1  ...   

   OCCUPATION_TYPE_Laborers  OCCUPATION_TYPE

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

1. Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
2. Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [549]:
# 1.
import numpy as np

# Step 1: Create a numpy array named y from the 'y' column, ensuring values are stored as integers
y = df_final['y'].astype(int).values

In [550]:
# 2: Create a numpy array named X from all the remaining features in df_final
X = df_final.drop(columns=['y']).values

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

1. Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 75% train and 25% test datasets (2.5 marks) 
    - Set random_state to 8 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
2. Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [551]:
# 1. 
from sklearn.model_selection import train_test_split

# Splitting the data into training (75%) and test (25%) sets with stratification
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=8, stratify=y
)

In [552]:
#2. 
from sklearn.preprocessing import StandardScaler

# Initializing the StandardScaler
scaler = StandardScaler()

# Fitting the scaler only on the training data and transforming it
X_train_scaled = scaler.fit_transform(X_train)

# Transforming the test data using the already fitted scaler
X_test_scaled = scaler.transform(X_test)

---
---

### Task 4. Logistic Regression and Random Forest Classifiers and Accuracies (Total Marks: 30)


**Question 9**. 

1. Train a Logistic Regression Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies   
2. Train a Random Forest Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies

When printing accuracies round the values to three decimal places.      

(Total: 10 marks)  

In [553]:
# 1. 

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize Logistic Regression with random_state set to 10
log_reg = LogisticRegression(random_state=10)

# Train the model
log_reg.fit(X_train_scaled, y_train)

# Predictions
y_train_pred_log = log_reg.predict(X_train_scaled)
y_test_pred_log = log_reg.predict(X_test_scaled)

# Calculate and print accuracies
train_accuracy_log = accuracy_score(y_train, y_train_pred_log)
test_accuracy_log = accuracy_score(y_test, y_test_pred_log)

print(f"Logistic Regression Train Accuracy: {round(train_accuracy_log, 3)}")
print(f"Logistic Regression Test Accuracy: {round(test_accuracy_log, 3)}")

Logistic Regression Train Accuracy: 0.657
Logistic Regression Test Accuracy: 0.658


In [554]:
#2
from sklearn.ensemble import RandomForestClassifier

# Initialize Random Forest with random_state set to 10
rf_clf = RandomForestClassifier(random_state=10)

# Train the model
rf_clf.fit(X_train_scaled, y_train)

# Predictions
y_train_pred_rf = rf_clf.predict(X_train_scaled)
y_test_pred_rf = rf_clf.predict(X_test_scaled)

# Calculate and print accuracies
train_accuracy_rf = accuracy_score(y_train, y_train_pred_rf)
test_accuracy_rf = accuracy_score(y_test, y_test_pred_rf)

print(f"Random Forest Train Accuracy: {round(train_accuracy_rf, 3)}")
print(f"Random Forest Test Accuracy: {round(test_accuracy_rf, 3)}")

Random Forest Train Accuracy: 0.977
Random Forest Test Accuracy: 0.898


**Question 10**. 

a) Comment and compare the training and test accuracies for each classifier computed in Question 9. What can we say about the extent of overfitting for each classifier? (5 marks)   
b) Comment and compare the accuracies across the two classifiers. Which classifier provides better forecasts? (5 marks)   
c) What can you say about the presence of nonlinearities in the dataset? (10 marks)   

(Total: 20 marks)  

# a) Comment and compare the training and test accuracies for each classifier. What can we say about the extent of overfitting for each classifier?
1. Logistic Regression:
Train Accuracy: 0.657
Test Accuracy: 0.658
The train and test accuracies are very close, indicating that the model is not overfitting. This suggests that Logistic Regression has a good generalization capability on the data.
Logistic Regression typically has similar training and test accuracies if the model is well-fitted and not overfitting, indicating that Logistic Regression is managing the complexity of the data well and not learning noise.

2. Random Forest:
Train Accuracy: 0.977
Test Accuracy: 0.898
The train accuracy is much higher than the test accuracy, indicating significant overfitting. The model fits the training data very well but struggles to generalize on the test data.
Random Forest often achieves very high training accuracy because it uses multiple decision trees, each capturing different patterns.
The test accuracy is significantly lower than the training accuracy, this indicates overfitting.
Overfitting: If the gap between training and test accuracy is large, it suggests that Random Forest is overfitting by capturing complex patterns in the training data that don't generalize well to unseen data.

# b) Comment and compare the accuracies across the two classifiers. Which classifier provides better forecasts?
Comparison:
Logistic Regression shows consistent but moderate performance on both training and test data.
Random Forest performs exceptionally well on training data but less so on test data due to overfitting.

Better Forecasts:
Despite overfitting, Random Forest has a higher test accuracy (0.898) compared to Logistic Regression (0.658), indicating that it provides better forecasts on the test data.

# c) What can you say about the presence of non-linearities in the dataset?

Nonlinearities:
Nonlinear relationships, occur when the effect of one variable on another is not constant, meaning the change is not proportional or predictable by a straight line. Nonlinear patterns can include curves, exponential growth, quadratic relationships, and other complex forms that cannot be captured by a simple straight line.
If Random Forest significantly outperforms Logistic Regression, especially on the test data, it indicates the presence of nonlinear relationships within the data. Random Forest’s ability to capture complex interactions and non-linear patterns often gives it an edge in such cases.

In the context of dataset, the Random Forest’s better performance compared to Logistic Regression suggests that there are nonlinearities present—patterns or relationships in the data that are not adequately captured by a linear model.



---
---